File: src/ovi_voice_assistant/memory/recall.py
"""Recall — multi-strategy retrieval with Reciprocal Rank Fusion."""

from __future__ import annotations

import logging
import math
from collections import defaultdict
from datetime import UTC, datetime
from enum import StrEnum

from ovi_voice_assistant.memory.embedder import Embedder
from ovi_voice_assistant.memory.store import MemoryStore
from ovi_voice_assistant.memory.types import Entity, Fact, RecallResult

logger = logging.getLogger(__name__)

# RRF constant — standard value from the literature.
_RRF_K = 60

# Temporal decay: e^(-days/HALF_LIFE). At 30 days, score ~0.37.
_TEMPORAL_HALF_LIFE = 30.0


class Budget(StrEnum):
    """Controls how many candidates to retrieve."""

    LOW = "low"  # 50 candidates — fast
    MID = "mid"  # 150 candidates — balanced
    HIGH = "high"  # 400 candidates — thorough


_BUDGET_LIMITS = {
    Budget.LOW: 50,
    Budget.MID: 150,
    Budget.HIGH: 400,
}


async def recall(
    *,
    bank_id: str,
    query: str,
    embedder: Embedder,
    store: MemoryStore,
    budget: Budget = Budget.MID,
    max_tokens: int = 2048,
) -> RecallResult:
    """Search memory using 4 parallel strategies + RRF fusion.

    Strategies:
        1. Semantic — cosine similarity on embeddings
        2. Keyword (BM25-lite) — LIKE query on fact text
        3. Entity graph — facts linked to entities mentioned in query
        4. Temporal — exponential decay favoring recent facts

    Results are merged via Reciprocal Rank Fusion and trimmed to token budget.
    """
    limit = _BUDGET_LIMITS.get(budget, 150)

    # 1. Semantic search
    try:
        query_emb = await embedder.embed_one(query)
    except Exception:
        logger.exception("Query embedding failed")
        query_emb = []

    semantic_results: list[tuple[str, float]] = []
    if query_emb:
        scored = store.search_facts_by_embedding(bank_id, query_emb, limit=limit)
        semantic_results = [(f.id, sim) for f, sim in scored]

    # 2. Keyword search
    keyword_facts = store.search_facts_by_text(bank_id, query, limit=limit)
    keyword_results = [(f.id, 1.0) for f in keyword_facts]

    # 3. Entity graph search
    entity_results = _entity_graph_search(bank_id, query, store, limit)

    # 4. Temporal search — score all facts by recency
    temporal_results = _temporal_search(bank_id, store, limit)

    # Merge via RRF
    rrf_scores = _rrf_merge(semantic_results, keyword_results, entity_results, temporal_results)

    if not rrf_scores:
        return RecallResult(results=[], total_candidates=0)

    # Rank by RRF score
    ranked_ids = sorted(rrf_scores, key=rrf_scores.get, reverse=True)

    # Fetch full facts and apply token budget
    all_facts = store.get_facts(bank_id, ranked_ids[:limit])
    fact_map = {f.id: f for f in all_facts}
    ranked_facts = [fact_map[fid] for fid in ranked_ids if fid in fact_map]

    # Set relevance scores
    for fact in ranked_facts:
        fact.relevance = rrf_scores.get(fact.id, 0.0)

    # Apply token budget
    budgeted = _apply_token_budget(ranked_facts, max_tokens)

    # Gather relevant entities
    entities = _gather_entities(bank_id, query, store)

    return RecallResult(
        results=budgeted,
        entities=entities,
        total_candidates=len(rrf_scores),
    )


def _entity_graph_search(
    bank_id: str, query: str, store: MemoryStore, limit: int
) -> list[tuple[str, float]]:
    """Find facts linked to entities mentioned in the query."""
    entities = store.get_entities(bank_id)
    if not entities:
        return []

    query_lower = query.lower()
    matching_entities = [e for e in entities if e.text.lower() in query_lower]

    results: list[tuple[str, float]] = []
    seen: set[str] = set()
    for entity in matching_entities:
        for fid in entity.fact_ids:
            if fid not in seen:
                seen.add(fid)
                results.append((fid, 1.0))
                if len(results) >= limit:
                    return results
    return results


def _temporal_search(
    bank_id: str, store: MemoryStore, limit: int
) -> list[tuple[str, float]]:
    """Score all facts by recency using exponential decay."""
    facts = store.get_facts(bank_id)
    if not facts:
        return []

    now = datetime.now(UTC)
    scored: list[tuple[str, float]] = []

    for fact in facts:
        if not fact.created_at:
            continue
        try:
            created = datetime.fromisoformat(fact.created_at)
            delta_days = (now - created).total_seconds() / 86400.0
            score = math.exp(-delta_days / _TEMPORAL_HALF_LIFE)
            scored.append((fact.id, score))
        except (ValueError, TypeError):
            continue

    scored.sort(key=lambda x: x[1], reverse=True)
    return scored[:limit]


def _rrf_merge(*ranked_lists: list[tuple[str, float]]) -> dict[str, float]:
    """Reciprocal Rank Fusion: score(d) = sum(1 / (k + rank_i(d)))."""
    scores: dict[str, float] = defaultdict(float)
    for ranked_list in ranked_lists:
        for rank, (fact_id, _) in enumerate(ranked_list):
            scores[fact_id] += 1.0 / (_RRF_K + rank + 1)
    return dict(scores)


def _apply_token_budget(facts: list[Fact], max_tokens: int) -> list[Fact]:
    """Keep facts until the token budget is exceeded (~4 chars/token)."""
    result: list[Fact] = []
    token_count = 0
    for fact in facts:
        fact_tokens = len(fact.text) // 4
        if token_count + fact_tokens > max_tokens:
            break
        result.append(fact)
        token_count += fact_tokens
    return result


def _gather_entities(bank_id: str, query: str, store: MemoryStore) -> list[Entity]:
    """Return entities mentioned in the query."""
    entities = store.get_entities(bank_id)
    query_lower = query.lower()
    return [e for e in entities if e.text.lower() in query_lower]

File: internal/memory/recall.go
package memory

import (
	"context"
	"log/slog"
	"math"
	"sort"
	"strings"
	"time"
)

// rrfK is the Reciprocal Rank Fusion constant — standard value from the literature.
const rrfK = 60

// temporalHalfLife: e^(-days/HALF_LIFE). At 30 days, score ~0.37.
const temporalHalfLife = 30.0

// Budget controls how many candidates to retrieve.
type Budget string

const (
	BudgetLow  Budget = "low"  // 50 candidates — fast
	BudgetMid  Budget = "mid"  // 150 candidates — balanced
	BudgetHigh Budget = "high" // 400 candidates — thorough
)

var budgetLimits = map[Budget]int{BudgetLow: 50, BudgetMid: 150, BudgetHigh: 400}

// rankedID is a fact id with a strategy score.
type rankedID struct {
	id    string
	score float64
}

// Recall searches memory using 4 parallel strategies + RRF fusion.
//
// Strategies:
//  1. Semantic — cosine similarity on embeddings
//  2. Keyword (BM25-lite) — LIKE query on fact text
//  3. Entity graph — facts linked to entities mentioned in query
//  4. Temporal — exponential decay favoring recent facts
//
// Results are merged via Reciprocal Rank Fusion and trimmed to token budget.
func Recall(ctx context.Context, bankID, query string, embedder Embedder, store *MemoryStore, budget Budget, maxTokens int) (RecallResult, error) {
	limit, ok := budgetLimits[budget]
	if !ok {
		limit = 150
	}
	if maxTokens <= 0 {
		maxTokens = 2048
	}

	// 1. Semantic search
	var semantic []rankedID
	queryEmb, err := embedder.EmbedOne(ctx, query)
	if err != nil {
		slog.Error("Query embedding failed", "err", err)
		queryEmb = nil
	}
	if len(queryEmb) > 0 {
		scored, err := store.SearchFactsByEmbedding(bankID, queryEmb, limit)
		if err != nil {
			return RecallResult{}, err
		}
		for _, s := range scored {
			semantic = append(semantic, rankedID{s.Fact.ID, s.Score})
		}
	}

	// 2. Keyword search
	keywordFacts, err := store.SearchFactsByText(bankID, query, limit)
	if err != nil {
		return RecallResult{}, err
	}
	keyword := make([]rankedID, 0, len(keywordFacts))
	for _, f := range keywordFacts {
		keyword = append(keyword, rankedID{f.ID, 1.0})
	}

	// 3. Entity graph search
	entities, err := store.GetEntities(bankID)
	if err != nil {
		return RecallResult{}, err
	}
	entityHits := entityGraphSearch(query, entities, limit)

	// 4. Temporal search
	allFacts, err := store.GetFacts(bankID, nil)
	if err != nil {
		return RecallResult{}, err
	}
	temporal := temporalSearch(allFacts, limit, time.Now().UTC())

	rrf := rrfMerge(semantic, keyword, entityHits, temporal)
	if len(rrf) == 0 {
		return RecallResult{Results: []Fact{}, TotalCandidates: 0}, nil
	}

	rankedIDs := make([]string, 0, len(rrf))
	for id := range rrf {
		rankedIDs = append(rankedIDs, id)
	}
	sort.SliceStable(rankedIDs, func(i, j int) bool {
		if rrf[rankedIDs[i]] == rrf[rankedIDs[j]] {
			return rankedIDs[i] < rankedIDs[j]
		}
		return rrf[rankedIDs[i]] > rrf[rankedIDs[j]]
	})

	fetch := rankedIDs
	if len(fetch) > limit {
		fetch = fetch[:limit]
	}
	facts, err := store.GetFacts(bankID, fetch)
	if err != nil {
		return RecallResult{}, err
	}
	factMap := make(map[string]Fact, len(facts))
	for _, f := range facts {
		factMap[f.ID] = f
	}
	ranked := make([]Fact, 0, len(fetch))
	for _, id := range rankedIDs {
		if f, ok := factMap[id]; ok {
			f.Relevance = rrf[id]
			ranked = append(ranked, f)
		}
	}

	return RecallResult{
		Results:         applyTokenBudget(ranked, maxTokens),
		Entities:        gatherEntities(query, entities),
		TotalCandidates: len(rrf),
	}, nil
}

// entityGraphSearch finds facts linked to entities mentioned in the query.
func entityGraphSearch(query string, entities []Entity, limit int) []rankedID {
	if len(entities) == 0 {
		return nil
	}
	q := strings.ToLower(query)
	var results []rankedID
	seen := map[string]bool{}
	for _, e := range entities {
		if !strings.Contains(q, strings.ToLower(e.Text)) {
			continue
		}
		for _, fid := range e.FactIDs {
			if seen[fid] {
				continue
			}
			seen[fid] = true
			results = append(results, rankedID{fid, 1.0})
			if len(results) >= limit {
				return results
			}
		}
	}
	return results
}

// temporalSearch scores all facts by recency using exponential decay.
func temporalSearch(facts []Fact, limit int, now time.Time) []rankedID {
	var scored []rankedID
	for _, f := range facts {
		if f.CreatedAt == "" {
			continue
		}
		created, err := parseTime(f.CreatedAt)
		if err != nil {
			continue
		}
		days := now.Sub(created).Hours() / 24.0
		scored = append(scored, rankedID{f.ID, math.Exp(-days / temporalHalfLife)})
	}
	sort.SliceStable(scored, func(i, j int) bool { return scored[i].score > scored[j].score })
	if len(scored) > limit {
		scored = scored[:limit]
	}
	return scored
}

func parseTime(s string) (time.Time, error) {
	for _, layout := range []string{time.RFC3339Nano, time.RFC3339, "2006-01-02T15:04:05", "2006-01-02"} {
		if t, err := time.Parse(layout, s); err == nil {
			return t, nil
		}
	}
	return time.Time{}, &time.ParseError{Value: s}
}

// rrfMerge implements Reciprocal Rank Fusion: score(d) = sum(1 / (k + rank_i(d))).
func rrfMerge(lists ...[]rankedID) map[string]float64 {
	scores := map[string]float64{}
	for _, list := range lists {
		for rank, item := range list {
			scores[item.id] += 1.0 / float64(rrfK+rank+1)
		}
	}
	return scores
}

// applyTokenBudget keeps facts until the token budget is exceeded (~4 chars/token).
func applyTokenBudget(facts []Fact, maxTokens int) []Fact {
	result := []Fact{}
	tokens := 0
	for _, f := range facts {
		ft := len(f.Text) / 4
		if tokens+ft > maxTokens {
			break
		}
		result = append(result, f)
		tokens += ft
	}
	return result
}

// gatherEntities returns entities mentioned in the query.
func gatherEntities(query string, entities []Entity) []Entity {
	q := strings.ToLower(query)
	out := []Entity{}
	for _, e := range entities {
		if strings.Contains(q, strings.ToLower(e.Text)) {
			out = append(out, e)
		}
	}
	return out
}
